display.py

```diff
@@ -21,271 +24,309 @@ def font(family='', bold=False, italic=False):
+        draw_text(self.ctx, text, center, **kw)
+
+    def draw_circle(self, radius=None, coords=None, **kw):
+        if radius is None:
+            radius = self.display.CELL_RADIUS * 2/3
+        center = self.display.convert_coords(coords)
+        draw_circle(self.ctx, center, radius, **kw)
+
+    def draw_line(self, coords0, coords1, stroke_width=2, color=(0, 0, 0, 1)):
+        self.ctx.set_source_rgba(*color)
+        self.ctx.set_line_width(convert_stroke_width(self.ctx, stroke_width))
+        self.ctx.move_to(*self.display.convert_coords(coords0))
+        self.ctx.line_to(*self.display.convert_coords(coords1))
         self.ctx.stroke()
 
-    def circle(self, size=None, *a, **kw):
-        # TODO: consistent name (circle vs. draw_circle)
-        if not size:
-            size = self.scale / 2.3
-        _draw_circle(self.ctx, self.mx, self.my, size, *a, **kw)
-
-    def line(self, x0, y0, x1, y1, size=1, color=(0, 0, 0, 1)):
-        r, g, b, a = color
-        self.ctx.set_source_rgba(b, g, r, a)
-        self.ctx.set_line_width(size)
-        self.ctx.move_to(self.x0 + self.scale*x0, self.y0 + self.scale*y0)
-        self.ctx.line_to(self.x0 + self.scale*x1, self.y0 + self.scale*y1)
-        self.ctx.stroke()
+    def draw_line_corners(self, c0, c1, *a, **kw):
+        self.draw_line(self.corners[c0], self.corners[c1], *a, **kw)
 
-def _draw_circle(ctx, x, y, size, color=(0, 0, 0, 1), fill=False):
-    r, g, b, a = color
-    ctx.set_source_rgba(b, g, r, a)
-    ctx.arc(x, y, size, 0, 6.3)
+def draw_text(ctx, text, coords, fontsize=12, family='', bold=False, italic=False, color=(0, 0, 0, 1)):
+    x, y = coords
+    ctx.set_font_size(convert_stroke_width(ctx, fontsize))
+    ctx.set_font_face(font(family, bold, italic))
+    ctx.set_source_rgba(*color)
+    _, _, w, h, dx, dy = ctx.text_extents(str(text))
+    ctx.move_to(x - w/2, y + h/2)
+    ctx.show_text(str(text))
+    ctx.stroke()
+
+def draw_circle(ctx, center, radius, color=(0, 0, 0, 1), fill=False, stroke_width=2):
+    x, y = center
+    ctx.set_source_rgba(*color)
+    ctx.arc(x, y, radius, 0, 6.3)
+    _fill_or_stroke(ctx, fill, stroke_width)
+
+def rotation_matrix_for_vector(vx, vy):
+    """
+    Returns a transformation matrix that converts the given vector into the vector (1, 0).
+    """
+    return cairo.Matrix(vx, vy, -vy, vx, 0, 0)
+
+def draw_polygon(ctx, points, color=(0, 0, 0, 1), fill=False, stroke_width=2):
+    """
+    Just a utility to make drawing polygons easier. I don't know why cairo doesn't have this.
+    You can also pass a grid-part-specific context.
+    """
+    if not isinstance(ctx, cairo.Context):
+        ctx = ctx.ctx
+    ctx.set_source_rgba(*color)
+    ctx.set_line_cap(cairo.LINE_CAP_BUTT)
+    ctx.set_line_join(cairo.LINE_JOIN_BEVEL)
+    ctx.move_to(*points[-1])
+    for point in points:
+        ctx.line_to(*point)
+    # go an extra step around so we get good bevels everywhere
+    ctx.line_to(*points[0])
```

Comment:
This is a nice touch. I would have supported adding this as a boolean parameter on my old `draw_rotated` context method.

Reply:
Yeah, I realized it was a problem when it originally had miters instead of bevels, and the human got drawn with one incredibly pointy side and one super lame side. MOST of the time it is so tiny that it doesn't matter, but that really stuck out.